core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala

```diff
@@ -0,0 +1,212 @@
+
+    logInfo("Running command: " + command)
+
+    // Mesos only work on linux platforms, so we assume bash is available is Mesos is used.
+    val pb = new ProcessBuilder("/bin/bash", "-c", command)
+
+    val currentEnvVars = pb.environment()
+    for (variable <- taskInfo.getExecutor.getCommand.getEnvironment.getVariablesList()) {
+      currentEnvVars.put(variable.getName, variable.getValue)
+    }
+
+    executorProc = pb.start()
+
+    new Thread("stderr reader for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        for (line <- Source.fromInputStream(executorProc.getErrorStream).getLines) {
+          System.err.println(line)
+        }
+      }
+    }.start()
+
+    new Thread("stdout reader for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        for (line <- Source.fromInputStream(executorProc.getInputStream).getLines) {
+          System.out.println(line)
+        }
+      }
+    }.start()
+
+    driver.sendStatusUpdate(TaskStatus.newBuilder()
+      .setState(TaskState.TASK_RUNNING)
+      .setTaskId(taskInfo.getTaskId)
+      .build)
+
+    new Thread("process waiter for mesos executor for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        executorProc.waitFor()
+        val (state, msg) = if (killed) {
+          (TaskState.TASK_KILLED, "")
+        }  else if (executorProc.exitValue() == 0) {
+          (TaskState.TASK_FINISHED, "")
+        } else {
+          (TaskState.TASK_FAILED, "Exited with status: " + executorProc.exitValue().toString)
+        }
+        // We leave the shuffle service running after the task.
+        cleanup(state, msg)
+      }
+    }.start()
+
+    taskId = taskInfo.getTaskId
+  }
+
+  override def error(d: ExecutorDriver, message: String) {
+    logError("Error from Mesos: " + message)
+  }
+
+  override def killTask(d: ExecutorDriver, t: TaskID) {
+    if (taskId == null) {
+      logError("Received killtask when no process is initialized")
+      return
+    }
+
+    if (!taskId.getValue.equals(t.getValue)) {
+      logError("Asked to kill task '" + t.getValue + "' but executor is running task '" +
+        taskId.getValue + "'")
+      return
+    }
+
+    assert(executorProc != null)
+    killed = true
+    // We only destroy the coarse grained executor but leave the shuffle
+    // service running for other tasks that might be reusing this executor.
+    // This is no-op if the process already finished.
+    executorProc.destroy()
+  }
+
+  def cleanup(state: TaskState, msg: String = ""): Unit = synchronized {
+    if (driver == null) {
+      logError("Cleaning up process but driver is not initialized")
+      return
+    }
+
+    if (executorProc == null) {
+      logDebug("Process is not started or already cleaned up")
+      return
+    }
+
+    assert(taskId != null)
+
+    driver.sendStatusUpdate(TaskStatus.newBuilder()
+      .setState(state)
+      .setMessage(msg)
```

Comment:
what happens if you set an empty message? Does it still get logged? Should we set the message only if it's not empty?